File: test/core/test_with_suffix.f90
program test_with_suffix

use, intrinsic :: iso_fortran_env, only: stderr=>error_unit
use filesystem

implicit none

valgrind : block

character(:), allocatable :: r

r = with_suffix("", ".h5")
if(r /= ".h5") error stop "with_suffix(,.h5) " // r // " /= .h5"

r = with_suffix("foo.h5", "")
if(r /= "foo") error stop "with_suffix(foo.h5,) " // r // " /= foo"

r = with_suffix(".foo.h5", ".txt")
if(r /= ".foo.txt") error stop "with_suffix(.foo.h5,.txt) " // r // " /= .foo.txt"

r = with_suffix(".h5", "")
if(r /= ".h5") error stop "with_suffix(.h5,) " // r // " /= .h5"

r = with_suffix(".h5", ".h5")
if(r /= ".h5.h5") then
  write(stderr,*) "ERROR: with_suffix(.h5,.h5) " // r
  error stop
end if

r = with_suffix('c:/a/hi.nc', '.h5')
if(r /= 'c:/a/hi.h5') then
  write(stderr,'(a)') "ERROR: with_suffix c:/a/hi.nc to .h5: " // r
  error stop
end if

r = with_suffix("my/file.h5", ".hdf5")
if (r /= "my/file.hdf5") error stop "with_suffix() failed: " // r

r = with_suffix("a/boo", ".h5")
if (r /= "a/boo.h5") error stop "with_suffix() failed: " // r

r = with_suffix("boo", ".h5")
if (r /= "boo.h5") error stop "with_suffix() failed: " // r

!> this is an odd case that would't be expected to be used
r = with_suffix("a/b/c.d/", ".hdf5")
if (r /= "a/b/c.d/.hdf5") error stop "with_suffix() failed: " // r

r = with_suffix("dir.h5/", ".hdf5")
if (r /= "dir.h5/.hdf5") error stop "with_suffix() failed: " // r

end block valgrind

print '(a)', "OK: with_suffix"

end program

File: test/core/test_with_suffix.cpp
#include <cstdlib>
#include <iostream>
#include <vector>
#include <string_view>
#include <tuple>

#include "ffilesystem.h"
#include "ffilesystem_test.h"


int main()
{

std::string r;
std::string ref;

int fail = 0;

const std::vector<std::tuple<std::string_view, std::string_view, std::string_view>> tests = {
  {"", ".h5", ".h5"},
  {"foo.h5", "", "foo"},
  {".foo.h5", ".txt", ".foo.txt"},
  {".h5", "", ".h5"},
  {".h5", ".h5", ".h5.h5"},
  {"a//b///c/", ".h5", "a//b///c/.h5"},
  {"c:/a/hi.nc", ".h5", "c:/a/hi.h5"},
  {"my/file.h5", ".hdf5", "my/file.hdf5"},
  {"a/boo", ".h5", "a/boo.h5"},
  {"boo", ".h5", "boo.h5"},
  {"a/b/c.d/", ".hdf5", "a/b/c.d/.hdf5"},
  {"dir.h5/", ".hdf5", "dir.h5/.hdf5"}
};

for (const auto& [input, suffix, expected] : tests) {
  r = fs_with_suffix(input, suffix);

  if (r != expected) {
    std::cerr << "FAIL: with_suffix(" << input << ", " << suffix << " = " + r + "   expected: " << expected << "\n";
    fail++;
  } else
    std::cout << "PASS: with_suffix(" << input << ", " << suffix << " = " + r + "\n";
}

  if(fail){
    std::cerr << "Failed: " << fail << " tests  backend: " << fs_backend() << "\n";
    return EXIT_FAILURE;
  }

  ok_msg("with_suffix C++");

  return EXIT_SUCCESS;
}
